converter_test.go

```diff
@@ -1299,30 +1599,135 @@ func TestSmallTimestampBinding(t *testing.T) {
 	})
 }
 
-func TestLargeTimestampBinding(t *testing.T) {
-	runSnowflakeConnTest(t, func(sct *SCTest) {
+func TestTimestampConversionWithoutArrowBatches(t *testing.T) {
+	timestamps := [3]string{
+		"2000-10-10 10:10:10.123456789", // neutral
+		"9999-12-12 23:59:59.123456789", // max
+		"0001-01-01 00:00:00.123456789"} // min
+	types := [3]string{"TIMESTAMP_NTZ", "TIMESTAMP_LTZ", "TIMESTAMP_TZ"}
+
+	runDBTest(t, func(sct *DBTest) {
 		ctx := context.Background()
-		timeValue, err := time.Parse("2006-01-02 15:04:05", "9000-10-10 10:10:10")
-		if err != nil {
-			t.Fatalf("failed to parse time: %v", err)
+
+		for _, tsStr := range timestamps {
+			ts, err := time.Parse("2006-01-02 15:04:05", tsStr)
+			if err != nil {
+				t.Fatalf("failed to parse time: %v", err)
+			}
+			for _, tp := range types {
+				for scale := 0; scale <= 9; scale++ {
+					t.Run(tp+"("+strconv.Itoa(scale)+")_"+tsStr, func(t *testing.T) {
+						query := fmt.Sprintf("SELECT '%s'::%s(%v)", tsStr, tp, scale)
+						rows := sct.mustQueryContext(ctx, query, nil)
+						defer rows.Close()
+
+						if rows.Next() {
+							var act time.Time
+							rows.Scan(&act)
+							exp := ts.Truncate(time.Duration(math.Pow10(9 - scale)))
+							if !exp.Equal(act) {
+								t.Fatalf("unexpected result. expected: %v, got: %v", exp, act)
+							}
+						} else {
+							t.Fatalf("failed to run query: %v", query)
+						}
+					})
+				}
+			}
 		}
-		parameters := []driver.NamedValue{
-			{Ordinal: 1, Value: DataTypeTimestampNtz},
-			{Ordinal: 2, Value: timeValue},
+	})
+}
+
+func TestTimestampConversionWithArrowBatchesFailsForDistantDates(t *testing.T) {
+	timestamps := [2]string{
+		"9999-12-12 23:59:59.123456789", // max
+		"0001-01-01 00:00:00.123456789"} // min
+	types := [3]string{"TIMESTAMP_NTZ", "TIMESTAMP_LTZ", "TIMESTAMP_TZ"}
+
+	expectedError := "Cannot convert timestamp"
+
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		ctx := WithArrowBatches(sct.sc.ctx)
+
+		pool := memory.NewCheckedAllocator(memory.DefaultAllocator)
+		defer pool.AssertSize(t, 0)
+		ctx = WithArrowAllocator(ctx, pool)
+
+		for _, tsStr := range timestamps {
+			for _, tp := range types {
+				for scale := 0; scale <= 9; scale++ {
+					t.Run(tp+"("+strconv.Itoa(scale)+")_"+tsStr, func(t *testing.T) {
+
+						query := fmt.Sprintf("SELECT '%s'::%s(%v)", tsStr, tp, scale)
+						_, err := sct.sc.QueryContext(ctx, query, []driver.NamedValue{})
+						if err != nil {
+							if !strings.Contains(err.Error(), expectedError) {
+								t.Fatalf("improper error, expected: %v, got: %v", expectedError, err.Error())
+							}
+						} else {
+							t.Fatalf("no error, expected: %v ", expectedError)
+
+						}
+					})
+				}
+			}
 		}
+	})
+}
 
-		rows := sct.mustQueryContext(ctx, "SELECT ?", parameters)
-		defer rows.Close()
+func TestTimestampConversionWithArrowBatchesAndWithOriginalTimestamp(t *testing.T) {
+	timestamps := [3]string{
```

Comment:
we have this list defined multiple timest in the tests?

Reply:
yes, but since it's not a large block of code, I think defining it in each test is more readable